torch/ao/quantization/_pt2e/prepare.py

```diff
@@ -0,0 +1,115 @@
+    _remove_output_observer,
+    _maybe_insert_observers_before_graph_output,
+    _save_state
+)
+from torch.fx import GraphModule
+
+from torch.ao.quantization import QConfigMapping
+from torch.ao.quantization.custom_config import PrepareCustomConfig
+from typing import Dict, Tuple, Any
+
+def prepare(
+        model: GraphModule,
+        quantizer: Quantizer,
+        is_qat: bool,
+        node_name_to_scope: Dict[str, Tuple[str, type]],
+) -> GraphModule:
+    model = quantizer.annotate(model)
+    quantizer.validate(model)
+
+    # Since we are mutating the graph as we go, we iterate over the original
+    # nodes before observer insertion, instead of model.graph.nodes.
+    nodes_before_observation = list(model.graph.nodes)
+
+    for node in nodes_before_observation:
+
+        if node.op == "placeholder":
+            pass
+        elif node.op in ("call_module", "call_method", "call_function"):
+            this_node_dtype_info = node.meta["target_dtype_info"]
+            if "val" in node.meta:
+                output_is_a_tensor = (
+                    this_node_dtype_info is not None and
+                    isinstance(node.meta["val"], FakeTensor)
+                )
+            else:
+                output_is_a_tensor = this_node_dtype_info is not None
+
+            skip_inserting_observers = (
+                not output_is_a_tensor
+            )
+
+            if skip_inserting_observers:
+                continue
+
+            _maybe_insert_input_observers_for_node(
+                node,
+                None, # qconfig
+                model,
+                named_modules,
+                model.graph,
+                None, # qhandler
+                PrepareCustomConfig(),
+                None, # backend_config
+            )
+
+            named_modules = dict(model.named_modules(remove_duplicate=False))
+            # this returns the new observer node if it was needed
+            maybe_output_obs_node = _maybe_insert_output_observer_for_node(node, model, named_modules, model.graph)
+
+            if maybe_output_obs_node is None:
+                continue
+            # Update users of original node to use the output observer
+            # instead. For example, change
+            #
+            #           next_node
+            #          /
+            #   cur_node -> obs
+            #
+            # to
+            #
+            #                 next_node
+            #                 /
+            #   cur_node -> obs
+            #
+            # We need to save orig users before updating uses because
+            # the list of users will change as we update uses
+            orig_users = list(node.users.keys())
+            for user_node in orig_users:
+                if user_node is maybe_output_obs_node:
+                    continue
+                user_node.replace_input_with(node, maybe_output_obs_node)
+            _is_observer_in_same_graph_ = _is_observer_in_same_graph(node, named_modules)
```

Comment:
Shouldnt all of this happen as part of `_maybe_insert_output_observer_for_node`? If not at least factor it out in a different function where you call `_maybe_insert_output_observer_for_node` and also update uses?

Reply:
makes sense, can I do this in a separate PR, I think we should be able to just move these things to `_maybe_insert_output_observer_for_node` and update all callsites